### PR TITLE
fix: 23762: Skip HDHM updates for leaves that are updated but not moved

### PR DESCRIPTION
**Pull request by artemananiev**
Fix summary:

* A new field is introduced to `VirtualLeafBytes`: path on disk
* This field is set to the current path when a leaf record is loaded from disk (actually, from anywhere, using `parseFrom()`)
* This field is used in `withPath()` method and preserved by `withValue()` methods
* This field is checked during flushes. If a leaf record is known to not be moved, the corresponding key/path mapping is not updated. This is the core of the optimization

Fixes: https://github.com/hiero-ledger/hiero-consensus-node/issues/23762
Signed-off-by: Artem Ananev <artem.ananev@swirldslabs.com>

### CODE REVIEW

File: platform-sdk/swirlds-virtualmap/src/test/java/com/swirlds/virtualmap/datasource/VirtualLeafBytesTest.java [hunk 1]
```diff
@@ -337,4 +338,29 @@ void writeToForHashing_withEmptyValueBytes() {
 
         assertArrayEquals(expected, actual, "hashing bytes should match");
     }
+
+    @Test
+    void isNewOrMovedIsTrueByDefault() {
```
Comment (imalygin) [inline platform-sdk/swirlds-virtualmap/src/test/java/com/swirlds/virtualmap/datasource/VirtualLeafBytesTest.java:345]: One more test is missing in which the leaf is neither moved nor new and `isNewOrMoved` return false

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java [hunk 2]
```diff
@@ -88,6 +114,15 @@ public long path() {
         return path;
     }
 
+    /**
+     * Indicates if this leaf record's path is different from where it was when loaded from
+     * disk. If the record was not loaded at all but created as new, the old path is set to
+     * -1, and this method still returns true.
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java]: Change -1 to Path.INVALID_PATH.

I believe we can have both `path` and `pathToDisk` to be `Path.INVALID_PATH` (a key is created and deleted). The method returns `false`, which would still be correct at where it's used (though as a deleted key it should not be in `dirtyLeaves`; we should not save the key to disk anyway) but I think this special case should be explicitly described and tested.

**Review by imalygin — DISMISSED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java [hunk 3]
```diff
@@ -137,15 +172,15 @@ public Bytes valueBytes() {
     }
 
     public VirtualLeafBytes<V> withPath(final long newPath) {
-        return new VirtualLeafBytes<>(newPath, keyBytes, value, valueCodec, valueBytes);
+        return new VirtualLeafBytes<>(newPath, path, keyBytes, value, valueCodec, valueBytes);
```
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java]: Minor: just double-checking: it's path, and not pathOnDisk that we use here, right? Note that in the two other with...() methods right below this one we use pathOnDisk for the second argument.

**Review by anthony-swirldslabs — APPROVED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java (hunk 3 above)
Comment (OlegMazurov) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java]: Shouldn't we use `pathOnDisk` instead of `path` here?

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java
Comment (OlegMazurov) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java:203]: Should we use `INVALID_PATH` instead of 0 here?

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/VirtualMap.java [hunk 4]
```diff
@@ -707,26 +707,25 @@ public void putBytes(@NonNull final Bytes keyBytes, @Nullable final Bytes valueB
...
+    @SuppressWarnings("unchecked")
     private <V> void put(final Bytes key, final V value, final Codec<V> valueCodec, final Bytes valueBytes) {
         throwIfImmutable();
         assert !isHashed() : "Cannot modify already hashed node";
         assert currentModifyingThreadRef.compareAndSet(null, Thread.currentThread());
         try {
             requireNonNull(key, NO_NULL_KEYS_ALLOWED_MESSAGE);
-            final long path = records.findPath(key);
-            if (path == INVALID_PATH) {
+            final VirtualLeafBytes<V> existing = records.findLeafRecord(key);
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/VirtualMap.java]: Can we preserve `records.findPath(key)` here for performance sake and use `records.findLeafRecord(path)` if `path` is found?

File: platform-sdk/swirlds-virtualmap/src/test/java/com/swirlds/virtualmap/datasource/VirtualLeafBytesTest.java (hunk 1 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/test/java/com/swirlds/virtualmap/datasource/VirtualLeafBytesTest.java:345]: A new unit test is added

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/VirtualMap.java [hunk 5]
```diff
@@ -789,6 +788,7 @@ public Bytes remove(@NonNull final Bytes key) {
                 final VirtualLeafBytes<?> lastLeaf = records.findLeafRecord(lastLeafPath);
                 assert lastLeaf != null;
                 cache.clearLeafPath(lastLeafPath);
+                // withPath() keeps track of the original path where the leaf was when loaded from disk
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/VirtualMap.java:793]: I'm not sure this comment referring to a now internal optimization is useful. It will not be properly maintained if that optimization changes or is removed.

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java (hunk 3 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java]: This is a good point. I think it should be `pathOnDisk`. This corner case doesn't affect current code, but it should be corrected anyway. I'll add a new unit test for it

Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java]: Yes, it's fixed now

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java:203]: No. Protobuf defaults should be used here, so it should be 0

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java (hunk 2 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java]: Changed to INVALID_PATH. Implementation is changed to return true, if both path and pathOnDisk are INVALID_PATH

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/VirtualMap.java (hunk 4 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/VirtualMap.java]: OK

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/VirtualMap.java (hunk 5 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/VirtualMap.java:793]: This optimization is internal to VirtualLeafBytes, but I still feel this comment is helpful. This is not a contract, but a clarification

**Review by anthony-swirldslabs — APPROVED** [review]

**Review by OlegMazurov — DISMISSED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java (hunk 2 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java]: I changed `isNewOrMoved()` JavaDoc, so it says this method shouldn't be used for leaf records with invalid paths. The corresponding assertion is added, too

**Review by anthony-swirldslabs — APPROVED** [review]
